### PR TITLE
Graph

### PR DESCRIPTION
**Pull request by kendricktang**
adding some graph algs.

### CODE REVIEW

**Comment by kendricktang** [comment]
No formal unit tests yet. I have some basic functionality demonstrated in the .py files themselves.

**Comment by kendricktang** [comment]
There are tests now.

File: graph_problems/graph_algs/circuitboard.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+def isdivisible(pins, wirecons):
+    """Pins is a set of pins. Wirecons must be a set of 2-tuples of pins.
+    Returns whether the pins can be divided onto two separate boards."""
+    mypins = {}
+    for pin in pins:
+        mypins[pin] = 0
+    wirecons = wirecons.copy()
```
Comment (port8080) [inline graph_problems/graph_algs/circuitboard.py]: comment why copy is required

File: graph_problems/graph_algs/circuitboard.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    for pin in mypins:
+        if not mypins[pin]:
+            # If pin hasn't been visited, arbitrarily mark it as on pcb 1.
+            mypins[pin] = 1
+
+        # Generate neighbors
+        neighbors = set()
+        for wirecon in wirecons:
+            if pin in wirecon:
+                neighbors.add(wirecon[pin == wirecon[0]])
```
Comment (port8080) [inline graph_problems/graph_algs/circuitboard.py]: comment on the cuteness in the indexing you are doing here

File: graph_problems/graph_algs/circuitboard.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
+def isdivisible(pins, wirecons):
+    """Pins is a set of pins. Wirecons must be a set of 2-tuples of pins.
```
Comment (port8080) [inline graph_problems/graph_algs/circuitboard.py]: a) Two board names should be constants - `BOARD_0 = -1` and `BOARD_1 = 1`
b) Prefer to create separate functions for legibility wherever you find yourself pulling a cool trick. Generally coolness correlates negatively with legibility
  - `wirecon[pin == wirecon[0]`
  - `mypins[otherpin] = -mypins[pin]`
  - `elif mypins[otherpin] + mypins[pin]`

**Comment by port8080** [comment]
awesomesauce, nits and then :shipit:

File: graph_problems/graph_algs/circuitboard.py (hunk 1 above)
Comment (kendricktang) [inline graph_problems/graph_algs/circuitboard.py]: I think I may have modified it in an earlier implementation and forgot to remove it.

**Pull request merged**